src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,936 @@
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "opto/rangeinference.hpp"
+#include "opto/type.hpp"
+#include "utilities/intn_t.hpp"
+#include "utilities/tuple.hpp"
+
+// If the cardinality of a TypeInt is below this threshold, use min widen, see
+// TypeIntPrototype<S, U>::normalize_widen
+constexpr juint SMALL_TYPEINT_THRESHOLD = 3;
+
+// This represents the result of an iterative calculation
+template <class T>
+class AdjustResult {
+public:
+  bool _progress; // whether there is progress compared to the last iteration
+  bool _present;  // whether the result is empty, typically due to the calculation arriving at contradiction
+  T _result;
+
+  bool empty() const {
+    return !_present;
+  }
+
+  static AdjustResult<T> make_empty() {
+    return {true, false, {}};
+  }
+};
+
+// This is the result of canonicalizing a simple interval (see TypeInt at
+// type.hpp)
+template <class U>
+class SimpleCanonicalResult {
+  static_assert(U(-1) > U(0), "bit info should be unsigned");
+public:
+  const bool _present;       // whether this is an empty set
+  const RangeInt<U> _bounds; // The bounds must be in the same half of the integer domain (see TypeInt)
+  const KnownBits<U> _bits;
+
+  SimpleCanonicalResult(bool present, const RangeInt<U>& bounds, const KnownBits<U>& bits)
+    : _present(present), _bounds(bounds), _bits(bits) {
+    if (!present) {
+      return;
+    }
+    // Do some verification
+    assert(bits.is_satisfied_by(bounds._lo) && bits.is_satisfied_by(bounds._hi), "must be canonical");
+    // 0b1000...
+    constexpr U mid_point = (std::numeric_limits<U>::max() >> 1) + U(1);
+    assert((bounds._lo < mid_point) == (bounds._hi < mid_point), "must be a simple interval");
+  }
+
+  bool empty() const {
+    return !_present;
+  }
+
+  static SimpleCanonicalResult<U> make_empty() {
+    return SimpleCanonicalResult(false, {}, {});
+  }
+};
+
+// Find the minimum value that is not less than lo and satisfies bits. If there
+// does not exist one such number, the calculation will overflow and return a
+// value < lo.
```

Comment:
I'm wondering if we should say anything more specific for this case. Maybe at least an example? It should probably not go here at the beginning, but somewhere further down.

Reply:
What do you mean? It is the doc for `adjust_lo`, of course it needs to be here, I think it is logical to say what the function does before saying how it does it. For that purpose I think it is very clear already and an example is not needed.

Reply:
@merykitty 
Hmm, I'm a little nervous about the case where `there does not exist one such number`. Because all your proof does is basically assume that there is such a `r`, and then shows that we compute it correctly.

But such proofs do not give us the guarantee that if there is no such `r`, that the computation indeed overflows, i.e. produces a number smaller than `lo`. That would be required for the correctness, no?

So I guess the proof / examples with overflow should happen further down, I'm just mentioning it up here because it is here that you say there can be an overflow. Hope that makes sense :see_no_evil:

Reply:
Thanks for the clarification, I have added a section for this case.

The fundamental logic here is that if there exists a result `r`, then the algorithm will find it. The opposite is also true, if the algorithm finds a satisfying `r`, then of course there exists one. This makes the negation of those statements also equivalent, the algorithm does not find a satisfying result if and only if there is no such value.

Reply:
Right, I think the missing part was that then the result is smaller than `lo`, which depends on that the result still satisfies bits... you now wrote a section about that, so that's better now :)